content/docs/userspace/gall-guide/11-fail.md

```diff
@@ -0,0 +1,139 @@
+`tang` will contain a stack trace.
+
+For `on-agent`, the `term` will be the head of the `sign` (`%poke-ack`, `%fact`,
+etc). The `tang` will contain a stack trace and a message of "closing
+subscription".
+
+For an `on-leave` failure, the `term` will always be `%leave`, and the `tang`
+will contain a stack trace.
+
+For a `%watch` failure, the `term` will be `%watch-not-unique`. The `tang` will
+include a message of "subscribe wire not unique", as well as the agent name, the
+`wire`, the target ship and the target agent.
+
+How you might handle these cases (if you wanted to manually handle them) depends
+on the logic of your particular agent.
+
+## Helper core
+
+Back in the lesson on lustar virtual arms, we briefly mentioned a common pattern
+is to define an alias of `hc` for a "helper core" like:
+
+```hoon
++*  this  .
+    def   ~(. (default-agent this %.n) bowl)
+    hc    ~(. +> bowl)
+```
+
+A helper core is a separate core composed into the subject of the agent core,
+containing useful functions for use by the agent arms. Such a helper core would
+typically contain functions that would only ever be used internally by the
+agent - more general functions would usually be included in a separate `/lib`
+library and imported with a [faslus](/docs/arvo/ford/ford#ford-runes) (`/+`)
+rune.
+
+The conventional pattern is to have the helper core _below_ the agent core, so the structure of the agent file is like:
+
+```
+[imports]
+[state types core]
+[agent core]
+[helper core]
+```
+
+Recall that the build system will implicitly compose any discrete expressions.
+If we simply added the helper core below the agent core, the agent core would be
+composed into the subject of the helper core, which is the opposite of what we
+want. Instead, we must inversely compose the two cores with a
+[tisgal](/docs/hoon/reference/rune/tis#-tisgal) (`=<`) rune. We add the tisgal
+rune directly above the agent core like:
+
+```hoon
+.....
+=<
+|_  =bowl:gall
++*  this      .
+    def   ~(. (default-agent this %|) bowl)
+    hc    ~(. +> bowl)
+++  on-init
+.....
+```
+
+We can then add the helper core below the agent core. The helper core is most
+typically a `door` like the agent core, also with the `bowl` as its sample. This
+is just so any functions you define in it have ready access to the `bowl`. It
+would look like:
+
+```hoon
+|_  =bowl:gall
+++  some-function  ...
+++  another  ....
+++  etc ...
+--
+```
+
+Back in the lustar virtual arm of the agent core, we give it an alias of `hc`
+and call it like so:
+
+```hoon
+hc  ~(. +>  bowl)
+```
+
+The composition will put the helper core in the tail of the tail of the agent
+core's subject, so we use a [censig](/docs/hoon/reference/rune/cen#-censig)
+expression to call `+>` of the subject (`.`) with the `bowl` as its sample.
+After that, any agent arms can make use of helper core functions by calling them
+like `(some-function:hc ....)`.
+
```

Comment:
I think it would also make sense to mention the use of `|^` inside of agent core arms as well. I think the convention is roughly that if the arm is only used by one of the agent arms, use `|^`. If you have an arm that multiple agent arms will use, put it in the helper core. Though I'd also say that any particularly heavy arm belongs in the helper core, even if it isn't used by more than one agent arm, but I don't know whether everybody does that.